packages/model/src/type-system/implementations/union.ts

```diff
@@ -86,19 +90,49 @@ class UnionTypeImpl<Ts extends types.Types> extends DefaultMethods<types.UnionTy
     value: unknown,
     decodingOptions?: decoding.Options,
   ): decoding.Result<types.Infer<types.UnionType<Ts>>> {
-    if (typeof value === 'object' && value) {
-      const object = value as Record<string, any>
-      const variantName = singleKeyFromObject(object)
-      if (variantName !== undefined && Object.keys(this.variants).includes(variantName)) {
-        return types
-          .concretise(this.variants[variantName])
-          .decodeWithoutValidation(object[variantName], decodingOptions)
-          .map((value) => Object.fromEntries([[variantName, value]]) as types.Infer<types.UnionType<Ts>>)
-          .mapError((errors) => prependVariantToAll(errors, variantName))
+    if (this.isTaggedUnion()) {
+      if (typeof value === 'object' && value) {
+        const object = value as Record<string, any>
+        const variantName = singleKeyFromObject(object)
+        if (variantName !== undefined && Object.keys(this.variants).includes(variantName)) {
+          return types
+            .concretise(this.variants[variantName])
+            .decodeWithoutValidation(object[variantName], decodingOptions)
+            .map((value) => Object.fromEntries([[variantName, value]]) as types.Infer<types.UnionType<Ts>>)
+            .mapError((errors) => prependVariantToAll(errors, variantName))
+        }
+      }
+      const prettyVariants = Object.keys(this.variants).join(' | ')
+      return decoding.fail(`union (${prettyVariants})`, value)
+    } else {
+      const errors: decoding.Error[] = []
+      let potentialDecoded: types.Infer<types.UnionType<Ts>> | null = null
+      for (const [variantName, variantType] of Object.entries(this.variants)) {
+        const concrete = types.concretise(variantType)
+        const result = concrete.decodeWithoutValidation(value, decodingOptions)
+        if (result.isOk) {
+          //look ahead with `validate` in order to get a correct variant if possible
+          const validateResult = concrete.validate(result.value as never)
```

Comment:
Why are you validating? It should decode without validation, when you validate the union you end up doing the same work twice

Reply:
Yes, this is necessary in order to infer the correct variant.
Take this example:
```typescript
 const model = types.union(
    {
      v1: types.number({ minimum: 0, maximum: 10 }),
      v2: types.number({ minimum: 20, maximum: 30 }),
    },
    { useTags: false },
  )
  const result = model.decode(25)
```
Without that validation it would decode (in the decodeWithoutValidation) as `{ v1: 25 }` and then fails the validation process. Meanwhile looking ahead with the `validate` will prevent this issue. It's less efficient but is the only way i find.
Do you have another suggestion?